File: tui/internal/dialog/movepicker_test.go
package dialog

import "testing"

func TestSortPickerQueues(t *testing.T) {
	cases := []struct {
		name        string
		sourceQueue string
		input       []string
		want        []string
	}{
		{
			name:        "DLQ source with matching non-DLQ queue pins it first",
			sourceQueue: "dlq.foo.bar",
			input:       []string{"alpha", "foo.bar", "zebra"},
			want:        []string{"foo.bar", "alpha", "zebra"},
		},
		{
			name:        "DLQ source case-insensitive prefix detection",
			sourceQueue: "DLQ.foo.bar",
			input:       []string{"foo.bar", "alpha"},
			want:        []string{"foo.bar", "alpha"},
		},
		{
			name:        "DLQ source candidate match is case-insensitive",
			sourceQueue: "dlq.Foo.Bar",
			input:       []string{"foo.bar", "alpha"},
			want:        []string{"foo.bar", "alpha"},
		},
		{
			name:        "DLQ source without matching candidate returns alphabetical",
			sourceQueue: "dlq.missing.queue",
			input:       []string{"zebra", "alpha", "mango"},
			want:        []string{"alpha", "mango", "zebra"},
		},
		{
			name:        "non-DLQ source returns alphabetical",
			sourceQueue: "foo.bar",
			input:       []string{"zebra", "alpha", "mango"},
			want:        []string{"alpha", "mango", "zebra"},
		},
		{
			name:        "empty list",
			sourceQueue: "dlq.foo",
			input:       []string{},
			want:        []string{},
		},
		{
			name:        "DLQ queues are de-prioritized after regular queues",
			sourceQueue: "foo.bar",
			input:       []string{"dlq.other", "regular", "another"},
			want:        []string{"another", "regular", "dlq.other"},
		},
		{
			name:        "system queues are de-prioritized last",
			sourceQueue: "foo.bar",
			input:       []string{"activemq.advisory", "regular", "statistics.foo"},
			want:        []string{"regular", "activemq.advisory", "statistics.foo"},
		},
		{
			name:        "full tier ordering: preferred, regular, DLQ, system",
			sourceQueue: "dlq.my.queue",
			input:       []string{"activemq.advisory", "dlq.other", "my.queue", "regular.a", "statistics.foo"},
			want:        []string{"my.queue", "regular.a", "dlq.other", "activemq.advisory", "statistics.foo"},
		},
		{
			name:        "IMQ source pins corresponding queue first",
			sourceQueue: "imq.foo.bar",
			input:       []string{"foo.bar", "alpha"},
			want:        []string{"foo.bar", "alpha"},
		},
		{
			name:        "IMQ queues are de-prioritized alongside DLQ queues",
			sourceQueue: "regular.queue",
			input:       []string{"imq.other", "regular.a", "dlq.other"},
			want:        []string{"regular.a", "dlq.other", "imq.other"},
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := sortPickerQueues(tc.sourceQueue, tc.input)
			if len(got) != len(tc.want) {
				t.Fatalf("sortPickerQueues() len = %d, want %d; got %v", len(got), len(tc.want), got)
			}
			for i := range got {
				if got[i] != tc.want[i] {
					t.Errorf("sortPickerQueues()[%d] = %q, want %q", i, got[i], tc.want[i])
				}
			}
		})
	}
}

File: mq-proxy-web/src/movepicker.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { tierForQueue, sortMoveTargets } from './movepicker.js';

test('tierForQueue: DLQ-stripped match against the source is preferred (tier 0)', () => {
  assert.equal(tierForQueue('foo.bar', 'dlq.foo.bar'), 0);
  assert.equal(tierForQueue('foo.bar', 'imq.foo.bar'), 0);
});

test('tierForQueue: an ordinary queue is regular (tier 1)', () => {
  assert.equal(tierForQueue('orders', 'dlq.foo.bar'), 1);
  assert.equal(tierForQueue('foo.bar', 'orders'), 1);
});

test('tierForQueue: dlq./imq.-prefixed queues are tier 2 unless they are the preferred match', () => {
  assert.equal(tierForQueue('dlq.other', 'orders'), 2);
  assert.equal(tierForQueue('imq.other', 'orders'), 2);
});

test('tierForQueue: activemq./statistics.-prefixed queues are tier 3 (bottom)', () => {
  assert.equal(tierForQueue('activemq.DLQ', 'orders'), 3);
  assert.equal(tierForQueue('statistics.broker', 'orders'), 3);
});

test('sortMoveTargets orders by tier then alphabetically, excludes the source queue', () => {
  const sorted = sortMoveTargets(
    ['zeta', 'activemq.DLQ', 'dlq.foo.bar', 'foo.bar', 'alpha', 'dlq.other'],
    'dlq.foo.bar',
  );
  assert.deepEqual(sorted, ['foo.bar', 'alpha', 'zeta', 'dlq.other', 'activemq.DLQ']);
});
